src/mlir/llvm.md

```diff
@@ -0,0 +1,178 @@
+ * "Native" types (integer, float) are still refering to a class (ex. U32) and
+   should be converted to appropriate LLVM types (ex. i32 with unsigned
+   semantics), and the subsequent operations will be converted to native LLVM
+   operations (ex. add).
+
+All other types should have been converted into a sequence of the types above and
+appropriate operations on them, including calling the runtime library.
+
+MLIR `module`s could have a class type (`!verona.class<...>`) which describes the
+fields it has, that will be lowered as LLVM struct types, for storage context.
+
+## Types
+
+Throughout the passes, pointer capabilities (such as `iso`, `mut` and `imm`)
+would have already been converted into compile-time checks (and failed if wrong),
+runtime checks (with calls to the runtime) or special lowering (sequence of
+instructions) on the code itself.
+
+This means by now we don't need those annotations and can drop all but the
+concrete types. Those would be concrete class or pointer/reference types to
+concrete classes, which would be indexed by name and we can just reuse the
+already declared class types via their LLVM struct types of the same name.
+
+We will probably need some mangling to get unique names across all classes and
+modules.
+
+### Regions / Cowns
+
+By now, all region information has already been converted to optimisations on
+the existing code, alias annotations on the remaining operations and calls to
+the runtime for scheduling, cleaning up, etc.
+
+There is nothing left to need special convertions.
+
+## Functions
+
+All functions, including class static and dynamic methods and compile-generated
+helpers, will be in the form of an MLIR function, with concrete typed arguments
+and full standard dialect control flow.
+
+Those can be directly lowered into the LLVM dialect on their own (via the
+dialects' own converters). Additional Verona dialect operations that exist will
+have their own implementation and be trivially converted.
+
+## Statements
+
+### new_region / new_object
+
+Those can be directly lowered to calls to the runtime to manipulate memory and
+control access and shouldn't have any special representation left in the
+Verona dialect.
+
+### call / static
+
+By now, all static calls have already been converted to a standard `call` to
+existing (mangled) functions and all dynamic calls have already been converted
+to `call_indirect` with the address calculated via vtable lookup.
+
+These lower to LLVM dialect directly and need no special convertion.
+
+### when
+
+`when` statements have already been converted to a sequence of creating the
+cown and calling the schedule runtime, which later on (see above) will be
+converted to standard calls, so again, nothing special here.
+
+### field_read / field_write
+
+All types at this stage are complete and concrete, which means reading/writing
+from/to fields in LLVM are just direct access to struct fields (ie. via GEP).
```

Comment:
We allow reading and writing fields in a union or interface type, which means we may not know the offset of the field. In this case, we need to look up the offset in the vtable (similar to a method call).

Reply:
In theory, union/interface field access constructs will not survive the MLIR pass, just like method calls, and will be replaced by vtable offset + load/store, and those lower directly. Whatever's left is on concrete types.

Reply:
I guess it doesn't really matter much, but I'm not sure I like treating operations on non-concrete types in a different place than operations on concrete types.

If we have a pass that lowers fields reads into a vtable load + field load, then we should also lower field read on concrete types. We just replace the first step of it by a constant value.

Reply:
> If we have a pass that lowers fields reads into a vtable load + field load, then we should also lower field read on concrete types. We just replace the first step of it by a constant value.

That's what it does, which means by the end, it's already a static call that have been lowered to a `std.call`.